fhir-persistence-jdbc/src/main/java/com/ibm/fhir/persistence/jdbc/util/JDBCQueryBuilder.java

```diff
@@ -609,14 +648,28 @@ protected SqlQueryData processChainedReferenceParm(QueryParameter queryParm) thr
             if (nextParameter != null) {
                 if (refParmIndex == 0) {
                     // Must build this first piece using px placeholder table alias, which will be replaced with a
-                    // generated value in the buildQuery() method.
-                    // Build this piece:P1.PARAMETER_NAME_ID = x AND (p1.STR_VALUE IN
+                    // generated value in the buildQuery() method. The CODE_SYSTEM_ID filter is added for issue #1366
+                    // due to the normalization of token values
+                    // Build this piece:P1.PARAMETER_NAME_ID = x AND AND P1.CODE_SYSTEM_ID = x AND (p1.TOKEN_VALUE IN
                     this.populateNameIdSubSegment(whereClauseSegment, currentParm.getCode(), PARAMETER_TABLE_ALIAS);
+
+                    // The resource type of the reference is encoded as the code system associated with the token value
+                    // so we need to add a filter to ensure we don't match logical-ids for other resource types
+                    // Note if the match is for any resource, we simply don't filter on the resource type
+                    final String codeSystemName = currentParm.getModifierResourceTypeName();
+                    if (codeSystemName != null && !codeSystemName.equals("*")) {
+                        Integer codeSystemId = identityCache.getCodeSystemId(codeSystemName);
+                        if (codeSystemId != null) {
+                            whereClauseSegment.append(AND).append(PARAMETER_TABLE_ALIAS + DOT).append(CODE_SYSTEM_ID).append(" = ")
+                                    .append(codeSystemId);
+                        }
+                    }
+                    
                     whereClauseSegment.append(AND);
                     whereClauseSegment.append(LEFT_PAREN);
-                    whereClauseSegment.append(PARAMETER_TABLE_ALIAS + DOT).append(STR_VALUE).append(IN);
+                    whereClauseSegment.append(PARAMETER_TABLE_ALIAS + DOT).append(TOKEN_VALUE).append(IN);
```

Comment:
shouldn't this be split into two appends? `.append(PARAMETER_TABLE_ALIAS).append(DOT)`

Reply:
yes, although this was how it was originally written...not sure why